spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/CloudVisionTemplate.java

```diff
@@ -0,0 +1,107 @@
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.protobuf.ByteString;
+
+import org.springframework.core.io.Resource;
+
+/**
+ * Spring Template offering convenience methods for interacting with the Cloud Vision
+ * APIs.
+ *
+ * @author Daniel Zou
+ *
+ * @since 1.1
+ */
+public class CloudVisionTemplate {
+
+	private final ImageAnnotatorClient imageAnnotatorClient;
+
+	public CloudVisionTemplate(ImageAnnotatorClient imageAnnotatorClient) {
+		this.imageAnnotatorClient = imageAnnotatorClient;
+	}
+
+	/**
+	 * Analyze an image and extract the feature of the image specified by {@code featureType}.
+	 *
+	 * <p>
+	 * A feature describes the kind of Cloud Vision analysis one wishes to perform on an
+	 * image, such as text detection, image labelling, facial detection, etc. A full list of
+	 * feature types can be found in {@link Feature.Type}.
+	 *
+	 * @param imageResource the image one wishes to analyze. The Cloud Vision APIs support
+	 *     image formats described here: https://cloud.google.com/vision/docs/supported-files
+	 * @param featureType the type of image analysis to perform on the image.
+	 * @return the results of image analysis
+	 */
+	public AnnotateImageResponse analyzeImage(Resource imageResource, Feature.Type featureType)
+			throws IOException {
+
+		return analyzeImage(imageResource, Collections.singleton(featureType));
+	}
+
+	/**
+	 * Analyze an image and extract the features of the image specified by
+	 * {@code featureTypes}.
+	 *
+	 * <p>
+	 * A feature describes the kind of Cloud Vision analysis one wishes to perform on an
+	 * image, such as text detection, image labelling, facial detection, etc. A full list of
+	 * feature types can be found in {@link Feature.Type}.
+	 *
+	 * @param imageResource the image one wishes to analyze. The Cloud Vision APIs support
+	 *     image formats described here: https://cloud.google.com/vision/docs/supported-files
+	 * @param featureTypes the types of image analysis to perform on the image.
+	 * @return the results of image analyses
+	 */
+	public AnnotateImageResponse analyzeImage(Resource imageResource, Set<Feature.Type> featureTypes)
+			throws IOException {
+
+		ByteString imgBytes = ByteString.readFrom(imageResource.getInputStream());
+		Image image = Image.newBuilder().setContent(imgBytes).build();
+
+		List<Feature> featureList = featureTypes.stream()
+				.map(featureType -> Feature.newBuilder().setType(featureType).build())
+				.collect(Collectors.toList());
+
+		BatchAnnotateImagesRequest request = BatchAnnotateImagesRequest.newBuilder()
+				.addRequests(
+						AnnotateImageRequest.newBuilder()
+								.addAllFeatures(featureList)
+								.setImage(image))
+				.build();
+
+		BatchAnnotateImagesResponse batchResponse = imageAnnotatorClient.batchAnnotateImages(request);
+		List<AnnotateImageResponse> annotateImageResponses = batchResponse.getResponsesList();
+
+		return annotateImageResponses.get(0);
```

Comment:
Why do you get only the first result? How about all others?
Can't that `getResponsesList()` be empty?

And what is the point to perform here a `BatchAnnotateImagesRequest` if we have only a single `image`?

Reply:
You probably do want to verify that there is only one, but that's what is expected because only one image was submitted in the batch.

Reply:
> Why do you get only the first result? How about all others? Can't that getResponsesList() be empty?

So this method will pass in exactly 1 image into the Cloud Vision APIs. Then, in the corresponding batch response, we are guaranteed to receive 1 response in the batch; there is a 1-to-1 correspondence between images you send to the service and your response objects you receive in the batch. If the image analysis fails for our image, we will still receive a batch of responses with size = 1; the response inside that batch will indicate the error that occurred for processing that image.

> And what is the point to perform here a BatchAnnotateImagesRequest if we have only a single image?

This is due to the design of the CloudVision APIs; it [does not offer a method for processing a Single image](https://cloud.google.com/vision/docs/reference/rpc/google.cloud.vision.v1#google.cloud.vision.v1.ImageAnnotator), only batch. From the perspective of their API, the single image request case collapses into the batch request case.

Reply:
Good. Can we consider then some much higher API instead of this low level `AnnotateImageResponse` and really throw an exception if that one failed?